core/network/src/discovery.rs

```diff
@@ -224,50 +237,58 @@ where
 		loop {
 			match self.kademlia.poll(params) {
 				Async::NotReady => break,
-				Async::Ready(NetworkBehaviourAction::GenerateEvent(ev)) => {
-					match ev {
-						KademliaOut::Discovered { .. } => {}
-						KademliaOut::KBucketAdded { peer_id, .. } => {
-							let ev = DiscoveryOut::Discovered(peer_id);
-							return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-						}
-						KademliaOut::FindNodeResult { key, closer_peers } => {
-							trace!(target: "sub-libp2p", "Libp2p => Query for {:?} yielded {:?} results",
-								key, closer_peers.len());
-							if closer_peers.is_empty() && self.num_connections != 0 {
-								warn!(target: "sub-libp2p", "Libp2p => Random Kademlia query has yielded empty \
-									results");
-							}
-						}
-						KademliaOut::GetValueResult(res) => {
-							let ev = match res {
-								GetValueResult::Found { results } => {
-									let results = results
-											.into_iter()
-											.map(|r| (r.key, r.value))
-											.collect();
-
-									DiscoveryOut::ValueFound(results)
-								}
-								GetValueResult::NotFound { key, .. } => {
-									DiscoveryOut::ValueNotFound(key)
-								}
-							};
-							return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-						}
-						KademliaOut::PutValueResult(res) => {
-							let ev = match res {
-								PutValueResult::Ok{ key, .. } =>  {
-									DiscoveryOut::ValuePut(key)
-								}
-								PutValueResult::Err { key, .. } => {
-									DiscoveryOut::ValuePutFailed(key)
+				Async::Ready(NetworkBehaviourAction::GenerateEvent(ev)) => match ev {
+					KademliaEvent::UnroutablePeer { peer, .. } => {
+						let ev = DiscoveryOut::UnroutablePeer(peer);
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					KademliaEvent::RoutingUpdated { peer, .. } => {
+						let ev = DiscoveryOut::Discovered(peer);
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					KademliaEvent::GetClosestPeersResult(res) => {
+						match res {
+							Err(e) => {
+								trace!(target: "sub-libp2p", "Kademlia query for closest peers failed: {:?}", e);
```

Comment:
The logging scheme is to mention what informed us.

```suggestion
								trace!(target: "sub-libp2p", "Libp2p => Kademlia query for closest peers failed: {:?}", e);
```

Also, in which situations can this happen? Isn't it better to use `error!`?

Reply:
> Also, in which situations can this happen? Isn't it better to use `error!`?

Right now, the only possible error for this query is a timeout. So what logging is desirable depends on the desired configuration / preferences. With the default timeout of 60 seconds, I would definitely not suggest logging these as errors, especially since these are "background queries". If the query timeout would be set large enough, I suppose a timeout could be considered an error. I'm not sure that is preferable though. Indeed, the timeout error also communicates the key and the peers found within the time constraints, so I suggest [this](https://github.com/paritytech/substrate/pull/3076/commits/f5259e24ee42c6c4b72510fe1ee51b11a713cae1#diff-7528f75916abf8807d5e5129430b9e19R316). What do you think?